### PR TITLE
Add `model_prompt` config param for LLMBlock

### LINKED ISSUE #136: Support a "blank" model prompt in `LLMBlock`
See aakankshaduggal/sdg#5

In order to support custom pipeline use cases where the pipeline author wishes to completely control the contents of the model prompt, we need some way of allowing the pipeline author to request a "blank template"

The POC was:

```
_BLANK_PROMPT = "{prompt}"

_MODEL_PROMPTS = {
    MODEL_FAMILY_MIXTRAL: _MODEL_PROMPT_MIXTRAL,
    MODEL_FAMILY_MERLINITE: _MODEL_PROMPT_MERLINITE,
    MODEL_FAMILY_BLANK: _BLANK_PROMPT
}
```

Put more simply, we want a pipeline author to be able to bypass the `model_prompt` here:

```
        prompts = [
            self.model_prompt.format(prompt=self._format_prompt(sample))
            for sample in samples
	]
```

How about `model_family: none` ?

```
  - name: custom_block
    type: LLMBlock
    config:
      config_path: ../../configs/knowledge/custom_config.yaml
      model_family: none
      gen_kwargs:
          model_id: custom_model
```

### PR DESCRIPTION
**Pull request by russellb**
Closes #136 

2363dfe docs: Add a doc to track pipeline config version history
87871d3 llmblock: add `model_prompt` config parameter
4f042ae Add test case for LLMBlock model_prompt parameter
7045456 llmblock: allow model_prompt=""

commit 2363dfe57dfcc2e08f1d8128e86edf6d78822f9d
Author: Russell Bryant <rbryant@redhat.com>
Date:   Mon Jul 15 14:35:26 2024 -0400

    docs: Add a doc to track pipeline config version history
    
    Our intiial pipeline configuration is version `1.0`. We haven't
    actually released anything yet, but I think it would be good to go
    ahead and get in the habit of managing this verison number as we make
    additions. Create a doc where we can keep a list of what changed at
    each version bump.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 87871d3d7b128fa12f7d073eb10b2f4401678115
Author: Russell Bryant <rbryant@redhat.com>
Date:   Mon Jul 15 14:40:18 2024 -0400

    llmblock: add `model_prompt` config parameter
    
    We previously always set the model prompt based on the model family.
    We have some cases where we'd like to override this default behavior
    explicitly as part of pipeline configuration. This parameter will do
    that.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 4f042ae88cf98c8e2335a7ab898622dadcb2c009
Author: Russell Bryant <rbryant@redhat.com>
Date:   Mon Jul 15 14:57:32 2024 -0400

    Add test case for LLMBlock model_prompt parameter
    
    Add a test case that ensures that passing `model_prompt` as an empty
    string results in `self.model_prompt` as empty. If this parameter is
    not specified (is None), `self.model_prompt` should be a non-empty
    string as a result of our default behavior of picking a prompt based
    on the model family.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

commit 7045456420c94dfb8b576e5d77777ababf73d233
Author: Mark McLoughlin <markmc@redhat.com>
Date:   Mon Jul 15 23:56:10 2024 +0100

    llmblock: allow model_prompt=""
    
    An anticipated use case for model_prompt is simply to disable
    this additional prompt. Currently the pipeline author would
    need to specify model_prompt="{prompt}" to achieve this.
    
    Make this easier by allowing model_prompt="" to have this
    meaning
    
    Signed-off-by: Mark McLoughlin <markmc@redhat.com>

### CODE REVIEW

**Comment by markmc** [comment]
I've added commit 2ff97ba to allow `model_prompt: ""` which is friendlier than `model_prompt: "{prompt}"`

File: docs/pipeline_config.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+## Pipeline Configuration Schema
+
+A schema for validating pipeline configuration can be found in [`src/instructlab/sdg/pipelines/schema/v1.json`](../src//instructlab/sdg/pipelines/schema/v1.json)
+
+## Version History
+
+| Version | Description |
+| ---     | --- |
+| 1.0     | Initial version |
+| 1.1     | Added `model_prompt` to `config` for `LLMBlock` |
```
Comment (markmc) [inline docs/pipeline_config.md:14]: FWIW, I'd be happy enough for us to keep changing 1.0 until we release it for a first time?

**Review by markmc — APPROVED** [review]

File: docs/pipeline_config.md (hunk 1 above)
Comment (russellb) [reply docs/pipeline_config.md:14]: Yeah, I'm fine with that.

I kinda got carried away thinking about how we'd document changes.  I'm totally fine pulling this from the PR for now and coming back to it later.

Comment (russellb) [reply docs/pipeline_config.md]: I rebased and pulled the version bump out. I left the doc in place though with a link to the schema and a version table that just lists 1.0.

**Review by russellb — COMMENTED** [review]
the changes @markmc added lgtm

File: docs/pipeline_config.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Pipeline Configuration
+
+Built-in pipeline configurations can be found in [`src/instructlab/sdg/pipelines/`](../src/instructlab/sdg/pipelines/).
+
+## Pipeline Configuration Schema
+
+A schema for validating pipeline configuration can be found in [`src/instructlab/sdg/pipelines/schema/v1.json`](../src//instructlab/sdg/pipelines/schema/v1.json)
```
[the remark this answers is not in the document]
Comment (russellb) [reply docs/pipeline_config.md:7]: oops
